File: packages/soulseek_protocol/lib/src/services/search_service.dart
import 'dart:async';

import '../connection/server_connection.dart';
import '../messages/buffer.dart';
import '../messages/codes.dart';
import '../messages/message.dart';

class SearchResult {
  final String username;
  final int ticket;
  final int freeUploadSlots;
  final int uploadSpeed;
  final int queueLength;
  final List<SearchResultFile> files;

  const SearchResult({
    required this.username,
    required this.ticket,
    required this.freeUploadSlots,
    required this.uploadSpeed,
    required this.queueLength,
    required this.files,
  });
}

class SearchService {
  final ServerTransport _server;
  final _resultsController = StreamController<SearchResult>.broadcast();
  int _nextTicket = 1;

  late final StreamSubscription _messageSub;

  SearchService({required ServerTransport server}) : _server = server;

  Stream<SearchResult> get results => _resultsController.stream;

  void init() {
    _messageSub = _server.messages.listen(_onMessage);
  }

  int search(String query) {
    final ticket = _nextTicket++;
    _server.sendMessage(SoulseekMessage(
      ServerCode.searchRequest,
      SearchRequest(query: query, ticket: ticket).serialize().toBytes(),
    ));
    return ticket;
  }

  void _onMessage(SoulseekMessage message) {
    if (message.code != ServerCode.searchResponse &&
        message.code != ServerCode.userSearchResponse) {
      return;
    }

    try {
      final buffer = ReadBuffer(message.payload);
      SearchResponseData response;
      try {
        response = SearchResponseData.parse(buffer);
      } catch (_) {
        response = SearchResponseData.parseOld(ReadBuffer(message.payload));
      }

      _resultsController.add(SearchResult(
        username: response.username,
        ticket: response.ticket,
        freeUploadSlots: response.freeUploadSlots,
        uploadSpeed: response.uploadSpeed,
        queueLength: response.queueLength,
        files: response.files,
      ));
    } catch (_) {
      // Skip malformed search responses
    }
  }

  void dispose() {
    _messageSub.cancel();
    _resultsController.close();
  }
}

File: Soulseek.Protocol/Services/SearchService.cs
using Soulseek.Protocol.Connection;
using Soulseek.Protocol.Messages;

namespace Soulseek.Protocol.Services;

public record SearchResult(
    string Username,
    int Ticket,
    int FreeUploadSlots,
    int UploadSpeed,
    int QueueLength,
    List<SearchResultFile> Files
);

public class SearchService
{
    private readonly IServerTransport _server;
    private readonly Subject<SearchResult> _resultsSubject = new();
    private int _nextTicket = 1;
    private IDisposable? _messageSub;

    public SearchService(IServerTransport server)
    {
        _server = server;
    }

    public IObservable<SearchResult> Results => _resultsSubject;

    public void Init()
    {
        _messageSub = _server.MessageStream.Subscribe(OnMessage);
    }

    public int Search(string query)
    {
        var ticket = _nextTicket++;
        _server.SendMessage(new SoulseekMessage(
            ServerCode.SearchRequest,
            new SearchRequest(query, ticket).Serialize().ToBytes()
        ));
        return ticket;
    }

    private void OnMessage(SoulseekMessage message)
    {
        if (message.Code != ServerCode.SearchResponse &&
            message.Code != ServerCode.UserSearchResponse)
            return;

        try
        {
            var buffer = new ReadBuffer(message.Payload);
            SearchResponseData response;
            try
            {
                response = SearchResponseData.Parse(buffer);
            }
            catch
            {
                response = SearchResponseData.ParseOld(new ReadBuffer(message.Payload));
            }

            _resultsSubject.OnNext(new SearchResult(
                response.Username,
                response.Ticket,
                response.FreeUploadSlots,
                response.UploadSpeed,
                response.QueueLength,
                response.Files
            ));
        }
        catch
        {
            // Skip malformed search responses
        }
    }

    public void Dispose()
    {
        _messageSub?.Dispose();
    }
}